src/compiler/compile/render_dom/index.ts

```diff
@@ -164,27 +164,41 @@ export default function dom(
 			`;
 		}
 
-		capture_state = (uses_props || writable_props.length > 0) ? x`
-			() => {
-				return { ${component.vars.filter(prop => prop.writable).map(prop => p`${prop.name}`)} };
-			}
-		` : x`
-			() => {
-				return {};
-			}
-		`;
+		// we need to filter out store subscriptions ($x) or $inject_state will try to call .set() on them, leading
+		// to a crash if store is not writable (and probably not intended behaviour to change store value)
+		const is_not_sub = (variable: Var) => variable.name.substr(0, 1) !== '$';
 
-		const writable_vars = component.vars.filter(variable => !variable.module && variable.writable);
-		inject_state = (uses_props || writable_vars.length > 0) ? x`
-			${$$props} => {
-				${uses_props && renderer.invalidate('$$props', x`$$props = @assign(@assign({}, $$props), $$new_props)`)}
-				${writable_vars.map(prop => b`
-					if ('${prop.name}' in $$props) ${renderer.invalidate(prop.name, x`${prop.name} = ${$$props}.${prop.name}`)};
-				`)}
-			}
-		` : x`
-			${$$props} => {}
-		`;
+		const capturable_vars = component.vars.filter(
+			variable => !variable.module && variable.writable && is_not_sub(variable)
+		);
+
+		if (uses_props || capturable_vars.length > 0) {
+
+			const capturable_props = writable_props.filter(is_not_sub);
+
+			const local_vars = capturable_vars.filter(variable => !variable.export_name);
+
+			const var_names = (variables: Var[]) => variables.map(prop => p`${prop.name}`);
+
+			capture_state = x`
+				({ props: $p = true, local: $l = true } = {}) => ({
```

Comment:
Was this left here on purpose? What I had suggested (and what it sounded like you agreed sounded good) was to have separate methods for capturing local state and for capturing props. What I especially don't want is something that returns both at once merged into one object, as that could lead to weird shadowing.

Reply:
Oops, sorry. I had remained focused on the previous discussion of `export { ... as ... }`. What I had understood was another method to deal with publicly visible (i.e. aliased) names, not to exclude props from `$capture_state` entirely.

My current implementation only uses internal names, both in what it returns and what it consumes (like the existing one, in fact). Is there really a risk of shadowing in this case? I don't really appreciate if this state could contain other things that could conflict, now or in the future...

If we had a `$capture_props` method, we'll need a `$inject_props` too, right? And this pair will have to deal with aliased vs internal names. Also, we need to sort the names of the methods out because what the existing `$capture_state` currently does is only exposing props as their internal names.

For now, I'm not sure HMR will ever need public names and, if so, I can workaround it with public API by using compiler's output. `svelte-devtools` also seems content with current `$inject_state`. So I was hoping efforts for `$capture_props` or so could be deferred until it was really needed.

Do you think I should split the methods right now? What should be the naming?

Reply:
Someone could (although it would be crazy) do something like

```js
let foo, bar;
export { foo as bar };
```

This is why I think props should be entirely excluded from the 'state' methods.

If you think we don't need `$capture_props`/`$inject_props` yet for tooling, that's fine, but I don't think `$capture_state` should have anything to do with props.

Reply:
To be clear, by this I don't mean that `$capture_state` should remove props from the object it returns, but just that it should disregard any `export`s and return an object of all of the top-level variables apart from those with `injected: true`.

Reply:
OK so, if I understand correctly, you're saying `$capture_state` should not take any argument and always return every writable variable that is neither injected, nor a store subscription.

That seems to make sense to me so I've updated the PR to do just that, and added a reactive statement in the test case to ensure that they don't end up in the captured state.